pkgs/by-name/ni/nixos-init/src/config.rs

```diff
@@ -1,43 +1,44 @@
-use std::env;
+use serde::Deserialize;
+use serde_json::from_reader;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::{Path, PathBuf};
 
 use anyhow::{Context, Result};
 
+#[derive(Deserialize)]
 pub struct Config {
     pub firmware: String,
     pub modprobe_binary: String,
     pub nix_store_mount_opts: Vec<String>,
     pub env_binary: Option<String>,
     pub sh_binary: Option<String>,
+    pub etc_basedir: Option<String>,
+    pub etc_metadata_image: Option<String>,
 }
 
 impl Config {
-    /// Read the config from the environment.
-    ///
-    /// These options are provided by wrapping the binary when assembling the toplevel.
-    pub fn from_env() -> Result<Self> {
-        let nix_store_mount_opts = required("NIX_STORE_MOUNT_OPTS")?
-            .split(',')
-            .map(std::borrow::ToOwned::to_owned)
-            .collect();
-
-        Ok(Self {
-            firmware: required("FIRMWARE")?,
-            modprobe_binary: required("MODPROBE_BINARY")?,
-            nix_store_mount_opts,
-            env_binary: optional("ENV_BINARY"),
-            sh_binary: optional("SH_BINARY"),
-        })
-    }
-}
+    /// Read the config from the metadata file in the toplevel directory.
+    pub fn from_toplevel(toplevel: impl AsRef<Path>, prefix: Option<&str>) -> Result<Self> {
+        let metadata_path = toplevel.as_ref().join("nixos-metadata.json");
 
-/// Read a required environment variable
-///
-/// Fail with useful context if the variable is not set in the environment.
-fn required(key: &str) -> Result<String> {
-    env::var(key).with_context(|| format!("Failed to read {key} from environment"))
-}
+        let metadata = if let Some(p) = prefix {
+            PathBuf::from(p).join(metadata_path.strip_prefix("/")?)
+        } else {
+            metadata_path
+        };
+
+        if !metadata
+            .try_exists()
+            .context("Failed to determine if metadata file exists, permissions issue?")?
+        {
+            anyhow::bail!("The metadata file '{}' does not exist", metadata.display());
+        }
 
-/// Read an optional environment variable
-fn optional(key: &str) -> Option<String> {
-    env::var(key).ok()
+        let reader = BufReader::new(File::open(metadata).context("Failed to open metadata file")?);
```

Comment:
Microoptimization:

Counterintuitively, reading JSON from a reader (even if buffered) is quite slow:

> Note that counter to intuition, this function is usually slower than reading a file completely into memory and then applying [from_str](https://docs.rs/serde_json/latest/serde_json/fn.from_str.html) or [from_slice](https://docs.rs/serde_json/latest/serde_json/fn.from_slice.html) on it. See [issue #160](https://github.com/serde-rs/json/issues/160).

From https://docs.rs/serde_json/latest/serde_json/fn.from_reader.html

Ideally, you'd call `fs::read` on the file and then `from_slice` on the resulting vec.